Add Serde get/set JSON functionality to Rust examples/docs (#615)
fermyon/developer: content/spin/rust-components.md

Spin has a key-value store built in. For information about using it from Rust, see [the key-value store tutorial](kv-store).

> If you want to go beyond just inserting values in a certain spot, Spin's Rust SDK provides an optional feature for using [Serde](https://docs.rs/serde/latest/serde). The Rust code below shows how to store and retrieve serializable objects from the key-value store:

Author: Yes, another good call. This page is long and with this `H3` we get some clarity and get to see it in the ToC.
Updated in latest commit.

```rust
use anyhow::Result;
use spin_sdk::{
    http::{Request, Response},
    http_component,
    key_value::{Store},
};
use serde::{Serialize, Deserialize};

#[http_component]
fn handle_request(_req: Request) -> Result<Response> {
    // Open the default key-value store
    let store = Store::open_default()?;
    // Define a User object using serde derive

>     #[derive(Serialize, Deserialize, Debug)]

Author: Removed `Debug`, thanks!

    struct User {
        fingerprint: String,
        location: String,
    }
    // Create an instance of a user object and populate the values

>     let user = User{fingerprint:"0x1234".to_owned(), location:"Brisbane".to_owned()};

Reviewer: nit: can we format these snippets with rustfmt?

Author: Sure think; sorry, I should have done that in the first place.
Formatting will be in the next commit.

    // Store the user object using the "my_json" key

>     store.set_json(String::from("my_json"), &user)?;

Reviewer: ```suggestion
    store.set_json("my_json", &user)?;
```

    // Retrieve the user object from the key-value store, using the "my_json" key
    let retrieved_user: User = store.get_json("my_json")?;
    // Respond to the request by revealing the user's fingerprint as the response body
    Ok(http::Response::builder().status(200).body(Some(retrieved_user.fingerprint.into()))?)
}
```

> The above example requires some additional configuration. Specifically, enabling the Spin SDK's optional `json` feature (from v1.2.0 onwards) and adding Serde's optional `derive` feature in the `dependencies` section of the application's `Cargo.toml` file:

Reviewer 2: The Rust templates don't reference `serde` by default so suggest rephrasing the second part, but I'm not sure if this is too wordy:

```suggestion
To use the `get_json` and `set_json` helpers, you must enable the Spin SDK's optional `json` feature in your `Cargo.toml` file.  To make your objects serialisable, you will also need a reference to `serde`; the example above also requires Serde's `derive` feature.  The relevant `Cargo.toml` entries look like this:
```

I wonder if it's worth adding a "details" expander along the lines of "Why are the helpers under an optional feature? // `get_json` and `set_json` depend on the `serde` and `serde_json` libraries. These are large libraries, so applications that depend on them are larger and take longer to build.  We didn't want users of the Rust SDK to incur these costs unless they actually wanted the feature."  Not ideal wording I fear but hopefully you get the idea!

> The above example requires some additional configuration. Specifically, enabling the Spin SDK's optional `json` feature (from v1.2.0 onwards) and adding Serde's optional `derive` feature in the `dependencies` section of the application's `Cargo.toml` file:

Author: Thanks, I have hoisted that up and it looks better. Good call.

### Serializing Objects to the Key-Value Store

The Spin key-value API stores and retrieves only lists of bytes. The Rust SDK provides helper functions that allow you to store and retrieve [Serde](https://docs.rs/serde/latest/serde) serializable values in a typed way. The underlying storage format is JSON (and is accessed via the `get_json` and `set_json` helpers). 

To use the `get_json` and `set_json` helpers, you must enable the Spin SDK's optional `json` feature in your `Cargo.toml` file. To make your objects serializable, you will also need a reference to `serde`. The relevant `Cargo.toml` entries look like this:

```
[dependencies]
// --snip --
serde = {version = "1.0.163", features = ["derive"]}
spin-sdk = {git = "https://github.com/fermyon/spin", version = "1.2.0", features = ["json"]}
// --snip --
```

The Rust code below shows how to store and retrieve serializable objects from the key-value store (note how the example below implements Serde's `derive` feature):

```rust
use anyhow::Result;
use serde::{Deserialize, Serialize};
use spin_sdk::{
    http::{Request, Response},
    http_component,
    key_value::Store,
};

> // Define a User object using serde derive

Reviewer 2: ```suggestion
// Define a serializable User type
```

#[derive(Serialize, Deserialize)]
struct User {
    fingerprint: String,
    location: String,
}

#[http_component]
fn handle_request(_req: Request) -> Result<Response> {
    // Open the default key-value store
    let store = Store::open_default()?;

    // Create an instance of a User object and populate the values
    let user = User {
        fingerprint: "0x1234".to_owned(),
        location: "Brisbane".to_owned(),
    };
    // Store the User object using the "my_json" key
    store.set_json("my_json", &user)?;
    // Retrieve the user object from the key-value store, using the "my_json" key
    let retrieved_user: User = store.get_json("my_json")?;
    // Return the user's fingerprint as the response body
    Ok(http::Response::builder()
        .status(200)
        .body(Some(retrieved_user.fingerprint.into()))?)
}
```

Once built and running (using `spin build` and `spin up`) you can test the above example in your browser (by visiting localhost:3000) or via curl, as shown below:

<!-- @selectiveCpy -->

```bash

> $ curl -i localhost:3000

Reviewer 2: Suggestion: 

```suggestion
$ curl localhost:3000
```

and remove the status/header lines from the output - they don't really show anything.
